### PR TITLE
Modules, logging and social updates

### LINKED ISSUE #84: Error log channel system

### LINKED ISSUE #121: Add better logging
+ Executed command logging
+ Joined guild
+ Left guild
+ Error logging if a command goes bork
+ etc...

### LINKED ISSUE #217: Marriage system
i d o n ' t  k n o w  h o w  t o  d o  i t  ,  b u t  t h e r e  i s  i t

### LINKED ISSUE #234: Integration with FOASS
[FOASS](http://foaas.com/) - Fuck Off As A Service

* Website: https://foaas.com/
* GitHub: https://github.com/tomdionysus/foaas

> FOAAS (Fuck Off As A Service) provides a modern, RESTful, scalable solution to the common problem of telling people to fuck off.

I request that this is implemented to switchblade because it provides many different ways to tell people to fuck off. It would be a switchblade-unique feature if FOASS was integrated, because no other discord bots have yet to use this service. Additionally, @pedrofracassi approved of such a great API.

image

-----

I noticed that, while searching around, there are two different tags that could be used:
* `:name` - name of who it is directed towards. however, not all messages from the list use `:name`, only certain ones do. 
* `:from` - name of who it is from. this tag should probably always remain the person who sent the command. personally, i believe this should just be omitted because switchblade already sends the person who used the command in the footer.

-----

Additionally, there are two filters that could be used. I've copied the descriptions from the website.
- shoutcloud - Integration with the ShoutCloud service. Please see shoutcloud.io for details.
  - You don't need this. You can probably do this with JS after getting the message.
- i18n - Translation provided by mymemory.translated.net. Specifying ?i18n on its own will attempt to detect your browser language, or you can specify the language to translate to using the ISO 639-1 language code, i.e. ?i18n=es
  - I mean, now you can swear in all the languages instead of just one. Why not?

-----

I propose that the command goes as follows:
* `s!foass` - sends a random message from the available list from FOASS.
* `s!foass list` - lists the possible messages to send from FOASS.
* `s!foass (message) (optional: name)` - chooses a specific message to send from FOASS to a certain user.
  * Example: `s!foass cool` would return `'How about a nice cup of shut the fuck up?` or
  * Example: `s!foass ing switchblade` would return `Fucking fuck off, switchblade.`
* For Filters: `s!foass (message) (optional: name) -f [shoutcloud|i18n]` - not sure how to do the arguments for this. ¯\_(ツ)_/¯

You should probably come up with a better arguments though, since in the future, `list` could be a possible message from FOASS.

### PR DESCRIPTION
**Pull request by bolsomito**
# Modules, logging and social updates
- [x] Set economy as module - `sendTo()`, `add()`, `remove()`, `checkBalance()`
- [x] Set guild config as module - `setConfig()`, `setLang()`
- [ ] Set social as module - `setPersonalText()`, `repUser()`, `marry()`, `setFavColor()`
- [ ] Add `s!personaltext <personal text>`, `s!rep <@user>`, `s!marry <@user>` and `s!favcolor <#hex>`
- [ ] Start logging stuff, probably a LoggingListener? - `logTransaction()`, `logGuildConfigChange()`, `logError()`, `logUserConfigChange()`

+ Closes #234 
+ Closes #217 
+ Closes #121 
+ Closes #84

### CODE REVIEW

File: src/commands/social/favcolor.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    )
+  }
+
+  async run ({ t, author, channel }, hexcolor) {
+    const embed = new SwitchbladeEmbed(author)
+    channel.startTyping()
+    // Regex
+    const REGEX_PATH = /^#(([0-9a-f]){3}){1,2}$/i
+    const hexcode = hexcolor.match(REGEX_PATH)
+    if(hexcode == null) {
```
Comment (bolsomito) [inline src/commands/social/favcolor.js]: standard here

File: src/commands/social/favcolor.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        embed.setColor(Constants.ERROR_COLOR)
+        .setTitle(t('commands:favcolor.invalidHex'))
+    } else {
+        // Database
+        const userData = await this.client.database.users.get(author.id)
+        const favColor = userData.favColor
+        userDoc.lastDaily = hexcode.input
```
Comment (bolsomito) [inline src/commands/social/favcolor.js]: vá alterar os user schemas no mongoose

**Comment by davipatury** [comment]
nice review

File: src/commands/social/favcolor.js (hunk 2 above)
Comment (Doges) [reply src/commands/social/favcolor.js]: eu já mudei, só que não enviei

File: src/utils/CanvasTemplates.js [hunk 3]
```diff
@@ -1,143 +1,148 @@
...
+    const DATABASE_QUERY = client.database.users.get(user.id)
+
+    const canvas = createCanvas(WIDTH, HEIGHT)
+    const ctx = canvas.getContext('2d')
+
+    const { lastDaily, money, personalText, favColor } = await DATABASE_QUERY
+
+    // Background gradient
+    const alphaToHex = (a) => Math.floor(a * 255).toString(16).padStart(2, '0')
+    const gradientColor = (a) => favColor + alphaToHex(a)
```
Comment (davipatury) [inline src/utils/CanvasTemplates.js]: This should use `Color` class.

**Comment by davipatury** [comment]
This PR needs to be splitted into smaller PRs.

File: src/modules/Rewards.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    if (now - lastBonus < LISTCORD_INTERVAL) {
+      return { ok: false, error: 'IN_INTERVAL', lastBonus, interval: LISTCORD_INTERVAL - (now - lastBonus) }
+    }
+
+    const { vote } = await this.validadeListcordVote({ user })
+    if (!vote || now - vote.lastVote >= LISTCORD_INTERVAL) {
+      return { ok: false, error: 'INVALID_VOTE' }
+    }
+
+    const value = this.contructor.LISTCORD_BONUS()
```
Comment (davipatury) [inline src/modules/Rewards.js]: `constructor`

**Comment by perronosaurio** [comment]
just_do_itttt
what are you waiting for this

**Comment by bolsomito** [comment]
Closing this, as this will be remade.

**Pull request closed without merging**